themis/spec/secure-cell.md

```diff
@@ -383,3 +383,186 @@ This value is stored as `40 0d 03 00` in little-endian encoding.
+
+### Example: symmetric keys
+
+With the following inputs (all encoded in ASCII):
+
+| Input          | Value |
+| -------------- | ----- |
+| encryption key | `au6aimoa8Pee8wahxi4Aique6eaxai2a` |
+| context data   | `additional context` |
+| plaintext      | `encrypted message` |
+
+Secure Cell in Seal mode produces the following output (encoded in base64):
+
+    AAEBQAwAAAAQAAAAEQAAAM5da3KkReYC7++OPbrI13UycoVi3s01Ji64WQ/KIe+3oF8cgLle19WC+tnaCg==
+
+which looks like this in hexadecimal (61 bytes):
+
+```
+00000000  00 01 01 40 0c 00 00 00  10 00 00 00 11 00 00 00  |...@............|
+00000010  ce 5d 6b 72 a4 45 e6 02  ef ef 8e 3d ba c8 d7 75  |.]kr.E.....=...u|
+00000020  32 72 85 62 de cd 35 26  2e b8 59 0f ca 21 ef b7  |2r.b..5&..Y..!..|
+00000030  a0 5f 1c 80 b9 5e d7 d5  82 fa d9 da 0a           |._...^.......|
+```
+
+There you can note the authentication token (44 bytes):
+
+```
+00000000  00 01 01 40 0c 00 00 00  10 00 00 00 11 00 00 00  |...@............|
+00000010  ce 5d 6b 72 a4 45 e6 02  ef ef 8e 3d ba c8 d7 75  |.]kr.E.....=...u|
+00000020  32 72 85 62 de cd 35 26  2e b8 59 0f -- -- -- --  |2r.b..5&..Y.    |
+00000030  -- -- -- -- -- -- -- --  -- -- -- --              |             |
+```
+
+and the actual encrypted data (17 bytes) in the end:
+
+```
+00000000  -- -- -- -- -- -- -- --  -- -- -- -- -- -- -- --  |                |
+00000010  -- -- -- -- -- -- -- --  -- -- -- -- -- -- -- --  |                |
+00000020  -- -- -- -- -- -- -- --  -- -- -- -- ca 21 ef b7  |            .!..|
+00000030  a0 5f 1c 80 b9 5e d7 d5  82 fa d9 da 0a           |._...^.......|
+```
+
+Since Seal mode has been used, the token and data are concatenated.
+In Token Protect mode they would have been returned as separate buffers.
+Here's how they look encoded in base64:
+
+    AAEBQAwAAAAQAAAAEQAAAM5da3KkReYC7++OPbrI13UycoVi3s01Ji64WQ8=
+
+    yiHvt6BfHIC5XtfVgvrZ2go=
+
+Try decrypting this data in Token Protect mode, it should work!
+
+Let's inspect the authentication token now.
+Match [the reference](#authentication-token--symmetric-keys)
+with the data as follows:
+
+| Field           | Offset | Data          | Meaning |
+| --------------- | ------ | ------------- | ------- |
+| algorithm ID    | 0x00   | `00 01 01 40` | `0x40010100` – AES-GCM-256, Soter KDF, PKCS#7 padding |
+| IV length       | 0x04   | `0c 00 00 00` | IV data is 12 bytes long |
+| auth tag length | 0x08   | `10 00 00 00` | auth tag is 16 bytes long |
+| message length  | 0x0C   | `11 00 00 00` | payload is 17 bytes long |
+| IV data         | 0x10   | `ce . . . 3d` | initialisation vector data |
+| auth tag        | 0x1C   | `ba . . . 0f` | authentication tag data |
+
+All the encoded values match the expectations.
+
+### Example: passphrases
+
+With the following inputs (all encoded in ASCII):
+
+| Input        | Value |
+| ------------ | ----- |
+| passphrase   | `au6aimoa8Pee8wahxi4Aique6eaxai2a` |
```

Comment:
`passphrase` here looks the same as `encryption key` in example above, is it intentional?

Also, what does ASCII encoding exactly mean? I guess every char corresponds to one byte in `encryption key`, 32 bytes in total, just like AES-256 expects. And here it is like 32-ASCII-char-string password. A little bit confusing imo.

Reply:
> passphrase here looks the same as encryption key in example above, is it intentional?

Yeah. I have been intentionally lazy to change it :) I guess I'll switch it for something more passphrase-looking, now that you've mentioned it.

> Also, what does ASCII encoding exactly mean?

Well, exactly what it says. You see *text* in the spec but cryptography works with bytes. There are various ways of converting them. In particular, the key might look like it's base32-encoded or something. For avoidance of doubt, I wanted to point out that those should be treated as ASCII strings. That is:

<table>
<tr>
<th>Text string</th><th>Encoded bytes</th>
</tr>
<tr>
<td>
<code>encrypted message</code>
</td>
<td>
<pre>
65 6e 63 72 79 70 74 65 64 20 6d 65 73 73 61 67
65
</pre>
</td>
</tr>
<tr>
<td>
<code>au6aimoa8Pee8wahxi4Aique6eaxai2a</code>
</td>
<td>
<pre>
61 75 36 61 69 6d 6f 61 38 50 65 65 38 77 61 68
78 69 34 41 69 71 75 65 36 65 61 78 61 69 32 61
</pre>
</td>
</tr>
</table>

Reply:
Or, to put in more practical way, these are the strings that you should give to command-line utilities to reproduce the results.

Reply:
@iamnotacake , nice catch!